### PR TITLE
Include error details into Identity exceptions

### LINKED ISSUE #5141: Identity should include the body of the bad response into the exception message
Somewhat real world scenario: an SDK developer did not set scope in the token request context.
Server returns 400 Bad Request. And the body says that the scope is empty which is the error.
But the exception being thrown just says "400 Bad Request". And you have to debug to see the actual error.

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #5141.

--
UWP CI is currently blocked until https://github.com/microsoft/vcpkg/issues/35279 is fixed / https://github.com/microsoft/vcpkg/pull/35116 is merged (#5181 would unblock)

### CODE REVIEW

File: sdk/identity/azure-identity/src/token_credential_impl.cpp [hunk 1]
```diff
@@ -117,11 +114,16 @@ AccessToken TokenCredentialImpl::GetToken(
         request = shouldRetry(statusCode, *response);
         if (request == nullptr)
         {
+          auto const bodyStream = response->ExtractBodyStream();
+          auto const bodyVec = bodyStream ? bodyStream->ReadToEnd(context) : response->GetBody();
+          auto const responseBody
+              = std::string(reinterpret_cast<char const*>(bodyVec.data()), bodyVec.size());
```
Comment (LarryOsterman) [inline sdk/identity/azure-identity/src/token_credential_impl.cpp:120]: reinterpret_cast? Is there a simpler cast?

**Review by LarryOsterman — APPROVED** [review]

File: sdk/identity/azure-identity/src/token_credential_impl.cpp (hunk 1 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/token_credential_impl.cpp:120]: Unfortunately, no:
```cpp
#include <cstdint>

int main() {
    return nullptr != static_cast<char const*>(static_cast<std::uint8_t const*>(nullptr));
}
```

```
<source>(4): error C2440: 'static_cast': cannot convert from 'const uint8_t *' to 'const char *'
<source>(4): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or parenthesized function-style cast
```
```
<source>:4:23: error: static_cast from 'const std::uint8_t *' (aka 'const unsigned char *') to 'const char *' is not allowed
    return nullptr != static_cast<char const*>(static_cast<std::uint8_t const*>(nullptr));
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```
<source>: In function 'int main()':
<source>:4:23: error: invalid 'static_cast' from type 'const uint8_t*' {aka 'const unsigned char*'} to type 'const char*'
    4 |     return nullptr != static_cast<char const*>(static_cast<std::uint8_t const*>(nullptr));
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

C-style case would work, but that is basically a `reinterpret_cast` anyway - yes, with simpler syntax, but I think we would hate it :)

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,8 @@
 
 ### Other Changes
 
+- [[#5141]](https://github.com/Azure/azure-sdk-for-cpp/issues/5141) Added error response details to the Identity exceptions thrown when the authority host returns error response.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/CHANGELOG.md:13]: nit: As a reader, I would wonder what are "Identity exceptions". Reword to use the actual exception class, i.e. `AuthenticationException`.

Comment (antkmsft) [reply sdk/identity/azure-identity/CHANGELOG.md:13]: #5211

**Pull request merged**